### PR TITLE
oci: Force kill the container process only if nothing else worked

### LINKED ISSUE #1327: waitContainerStop() should SIGKILL the container process only as last option
I was looking into [StopContainer()](https://github.com/kubernetes-incubator/cri-o/blob/master/oci/oci.go#L608-L635) from `oci/oci.go`, and I realized that [waitContainerStop()](https://github.com/kubernetes-incubator/cri-o/blob/master/oci/oci.go#L564-L605) was actually doing more than just waiting for the process to terminate. Indeed, if it reaches the timeout, it kills directly the container PID by issuing a signal to it.
I have the feeling this is wrong, or at least this should happen only after we tried both [a regular kill](https://github.com/kubernetes-incubator/cri-o/blob/master/oci/oci.go#L620) and a [kill SIGKILL](https://github.com/kubernetes-incubator/cri-o/blob/master/oci/oci.go#L630). We should call a simplified version of the current `waitContainerStop()` [here](https://github.com/kubernetes-incubator/cri-o/blob/master/oci/oci.go#L623) and the regular version doing the SIGKILL as the last thing we can try [here](https://github.com/kubernetes-incubator/cri-o/blob/master/oci/oci.go#L634).

### PR DESCRIPTION
**Pull request by sboeuf**
The function waitContainerStop() in oci/oci.go was designed to check
on a container process and wait for it until it returns, or returns
an error if a timeout was reached.

Now, looking at the way this function is used, it is called twice
from the same function StopContainer() with the idea of trying to
wait the process properly first. Then, if the process is not
terminating as expected, another signal is sent to the container
process, and another call to waitContainerStop() is issued.

This second call should be the only one calling into SIGKILL because
this is the last chance to stop the container process. But from the
first call into this function, we should only retrieve sort of a
boolean value telling if the process is still around, instead of
trying to kill it with a SIGKILL already.

This commit modifies waitContainerStop() by adding a simple flag
indicating if this function should try to force kill the container
process using SIGKILL signal in case the process is still around
after the timeout has been reached.

Fixes #1327

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

### CODE REVIEW

**Comment by sboeuf** [comment]
@mrunalp @runcom PTAL

**Comment by wking** [comment]
> This second call should be the only one calling into SIGKILL because
this is the last chance to stop the container process. But from the
first call into this function, we should only retrieve sort of a
boolean value telling if the process is still around, instead of
trying to kill it with a SIGKILL already.

It seems to me like we need three phases to be safe:

1. If `timeout > 0`, [call the runtime's `kill` with a gentle signal][1].  Wait `timeout` to see if that worked.  If not…
2. [Call the runtime's `kill` with `KILL`][2].  Wait `killContainerTimeout` to see if that worked.  If not…
3. If we have a container PID (it's [only required on Linux][3]), [send a `KILL` directly][4].  Wait `killContainerTimeout` again to see if that worked.

As of 2d1658f02, you're missing the wait portion of step 3, which means things that get that far (unlikely, but could happen with an uninterruptible disk sleep) may be [marked “finished”][5] before the container process actually exits.  If that analysis makes sense, we can drop the the `unix.Kill` entirely from `waitContainerStop` and instead call it directly from `StopContainer` before launching another `waitContainerStop` to judge it's success.  Something like:

```go
// StopContainer stops a container. Timeout is given in seconds.
func (r *Runtime) StopContainer(ctx context.Context, c *Container, timeout int64) error {
	c.opLock.Lock()
	defer c.opLock.Unlock()

	// Check if the process is around before sending a signal
	err := unix.Kill(c.state.Pid, 0)
	if err == unix.ESRCH {
		c.state.Finished = time.Now()
		return nil
	}

	if timeout > 0 {
		if err = utils.ExecCmdWithStdStreams(os.Stdin, os.Stdout, os.Stderr, r.Path(c), "kill", c.id, c.GetStopSignal()); err != nil {
			return fmt.Errorf("failed to stop container %s, %v", c.id, err)
		}
		err = waitContainerStop(ctx, c, time.Duration(timeout)*time.Second)
		if err == nil {
			return nil
		}
		logrus.Warnf("stop container %q via the runtime timed out: %v", c.ID(), err)
	}

	if err := utils.ExecCmdWithStdStreams(os.Stdin, os.Stdout, os.Stderr, r.Path(c), "kill", "--all", c.id, "KILL"); err != nil {
		return fmt.Errorf("failed to stop container %s, %v", c.id, err)
	}
	err = waitContainerStop(ctx, c, killContainerTimeout)
	if err == nil {
		return nil
	}
	logrus.Warnf("kill container %q via the runtime timed out: %v", c.ID(), err)

	err := unix.Kill(c.state.Pid, unix.SIGKILL)
	if err == nil || err == unix.ESRCH {
		return nil
	}
	if err != nil {
		return fmt.Errorf("failed to kill process: %v", err)
	}
	err = waitContainerStop(ctx, c, killContainerTimeout)
	if err == nil {
		return nil
	}
	logrus.Warnf("kill container %q directly via PID %d timed out: %v", c.ID(), c.state.Pid, err)
	return err
}
```

On an unrelated note, `waitContainerStop`'s `timeout` argument is superfluous.  This is what [`WithTimeout`][6] is for, and [`waitContainerStop` is already monitoring its `ctx`][7] to track that sort of thing.  I don't know if we want to address that as part of this PR or punt it to follow-up work.

[1]: https://github.com/kubernetes-incubator/cri-o/blob/0e0c820f0c0a4220a0462d2b89bd18a9f0012e2f/oci/oci.go#L670
[2]: https://github.com/kubernetes-incubator/cri-o/blob/0e0c820f0c0a4220a0462d2b89bd18a9f0012e2f/oci/oci.go#L680
[3]: https://github.com/opencontainers/runtime-spec/blame/v1.0.1/runtime.md#L25
[4]: https://github.com/kubernetes-incubator/cri-o/blob/0e0c820f0c0a4220a0462d2b89bd18a9f0012e2f/oci/oci.go#L597
[5]: https://github.com/kubernetes-incubator/cri-o/blob/0e0c820f0c0a4220a0462d2b89bd18a9f0012e2f/oci/oci.go#L603
[6]: https://godoc.org/golang.org/x/net/context#WithTimeout
[7]: https://github.com/kubernetes-incubator/cri-o/blob/0e0c820f0c0a4220a0462d2b89bd18a9f0012e2f/oci/oci.go#L592-L594

**Comment by sboeuf** [comment]
@wking what you're detailing here kind of makes sense, but I think this could be done in two steps, and this patch is the first step, making sure we don't call SIGKILL twice, which does not make sense.

**Comment by wking** [comment]
On Tue, Mar 06, 2018 at 09:35:36PM +0000, Sebastien Boeuf wrote:
> @wking what you're detailing here kind of makes sense, but I think
> this could be done in two steps…

Sure.  If I have time, I can work up a PR for my proposal, but I don't
mind if I do that before or after this PR lands.

**Comment by sboeuf** [comment]
@rhatdan could you trigger a retest for `integration_rhel` please, this is weird that only this one failed here.

**Comment by wking** [comment]
[Failure][1] is #1424.

[1]: https://ci.openshift.redhat.com/jenkins/job/test_pull_request_crio_integration_rhel/596/s3/download/gathered/artifacts/testout.txt

**Comment by rhatdan** [comment]
Should we turn the test that is failing off on RHEL?  
/test integration_rhel

**Comment by rhatdan** [comment]
LGTM
@mrunalp @runcom Got this to pass can you guys review and Merge?

File: oci/oci.go [hunk 1]
```diff
@@ -561,7 +561,7 @@ func (r *Runtime) UpdateContainer(c *Container, res *rspec.LinuxResources) error
 	return nil
 }
 
-func waitContainerStop(ctx context.Context, c *Container, timeout time.Duration) error {
+func waitContainerStop(ctx context.Context, c *Container, timeout time.Duration, forceKill bool) error {
```
Comment (runcom) [inline oci/oci.go]: PR looks ok, but afaict it's not actually force killing anything right? maybe this should be ignoreKill (just nitpicking)

Comment (sboeuf) [reply oci/oci.go]: Sure, I can change the name here.

Comment (sboeuf) [reply oci/oci.go]: Done !

**Comment by sboeuf** [comment]
@runcom @rhatdan please re-trigger the tests

**Comment by runcom** [comment]
LGTM

**Comment by rhatdan** [comment]
All Green. Merging.

**Pull request merged**